src/win32/posix_w32.c

```diff
@@ -981,3 +981,73 @@ int p_inet_pton(int af, const char *src, void *dst)
 	errno = EINVAL;
 	return -1;
 }
+
+ssize_t p_pread(git_file fd, void *data, size_t size, git_off_t offset)
+{
+	size_t mmap_alignment;
+	size_t page_offset;
+	git_off_t page_start;
+	unsigned char *map_data;
+	git_map map;
+	int error;
+
+	assert(data && size);
+
+	if (!git__is_ssizet(size)) {
+		errno = EINVAL;
+		SetLastError(ERROR_INVALID_PARAMETER);
+		return -1;
+	}
+
+	if ((error = git__mmap_alignment(&mmap_alignment)) < 0)
+		return error;
+
+	/* the offset needs to be at the mmap boundary for the platform */
+	page_offset = offset % mmap_alignment;
+	page_start = offset - page_offset;
+
+	if ((error = p_mmap(&map, page_offset + size, GIT_PROT_READ,
+					GIT_MAP_SHARED, fd, page_start)) < 0)
+		return error;
+
+	map_data = (unsigned char *)map.data;
+	memcpy(data, map_data + page_offset, size);
+	p_munmap(&map);
+
+	return size;
+}
+
+ssize_t p_pwrite(git_file fd, const void *data, size_t size, git_off_t offset)
+{
+	size_t mmap_alignment;
+	size_t page_offset;
+	git_off_t page_start;
+	unsigned char *map_data;
+	git_map map;
+	int error;
+
+	assert(data && size);
+
+	if (!git__is_ssizet(size)) {
+		errno = EINVAL;
+		SetLastError(ERROR_INVALID_PARAMETER);
+		return -1;
+	}
+
+	if ((error = git__mmap_alignment(&mmap_alignment)) < 0)
+		return error;
+
+	/* the offset needs to be at the mmap boundary for the platform */
+	page_offset = offset % mmap_alignment;
+	page_start = offset - page_offset;
+
+	if ((error = p_mmap(&map, page_offset + size, GIT_PROT_WRITE,
```

Comment:
If you try to build with `NO_MMAP` on windows, we'll call `p_pwrite` which calls `mmap`, which seems to violate the user's request and expectations.  Let's not implement `p_pwrite` and `p_pread` on Windows, let's just disable the ability to build with `NO_MMAP` on Windows so that these don't get called and then not provide a `p_pwrite` or `p_pread` implementation in win32 at all.  If somebody wants to fix the `NO_MMAP` situation, they can provide a Windows-specific implementation of `p_pread`/`p_pwrite`.

Reply:
I realized preadv/pwritev does not implement scatter/gather IO. I prefer to rework this PR.

Since this PR is built on top of: https://github.com/libgit2/libgit2/pull/5583, I request @ethomson to please review that instead. I have implemented pread/pwrite for win32 using Win32 overlapped file IO.